File: JumpGame.java
class Solution {
    public boolean canJump(int[] nums) {
        int goal = nums.length - 1;

        for (int i = goal - 1; i > -1; i--){
            if (nums[i] + i >= goal){
                goal = i;
            }
        }

        if (goal == 0){
            return true;
        }
        return false;
    }
}

// Time: O(n)
// Space: O(1)

File: JumpGame.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        goal = len(nums) - 1

        for i in range(len(nums)-2, -1, -1): #(start, end, jump)
            if nums[i] + i >= goal:
                goal = i
        
        return goal == 0

    #Time: O(nums)
    #Space: O(1)
